src/PIL/ImageGrab.py

```diff
@@ -140,7 +140,33 @@ def grabclipboard():
                 return BmpImagePlugin.DibImageFile(data)
         return None
     else:
-        if shutil.which("wl-paste"):
+        if shutil.which("loginctl"):
+            try:
+                loginctl = subprocess.check_output("loginctl").decode().split("\n")
+            except subprocess.CalledProcessError:
+                loginctl = None
+        else:
+            loginctl = None
+
+        if loginctl is not None:
+            username = subprocess.check_output("whoami").decode().strip("\n")
+            sessionid = [
+                line.split()[0] for line in loginctl if username in line.split()
+            ][0]
```

Comment:
Would it be simpler to just use [`os.getsid()`](https://docs.python.org/3/library/os.html#os.getsid)?

Reply:
I'm not sure what pid to pass to the function though, `os.getsid(0)` (sid of current process) seems to be different from the one from `loginctl`

Reply:
Alternatively, the session type can be obtained by checking the environment variable `XDG_SESSION_TYPE`, but I didn't go with that route because user can override it (security reason?). Or maybe it is fine? That would simplify a lot.

Reply:
I don't think the user being able to control whether xclip or wl-paste is used constitutes a security risk.

It's just hard to see a consensus on StackOverflow about which methods work and which ones don't. An alternative method is suggested at https://unix.stackexchange.com/a/559950, of checking `WAYLAND_DISPLAY` instead. For some reason, when we added testing for `wl-paste` in #7094, we set `WAYLAND_DISPLAY`, but not `XDG_SESSION_TYPE` - so at the moment, only the first variable is available in our CI jobs. Something to keep in mind.

Reply:
Indeed there doesn't seem to have universal method for the check. Should I go ahead and change to using environment variable?